### ISSUE TITLE
Prevent errors due to unique data stream name constraint

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#6727]
## Feature Description

GA4 requires that all data stream names for a property are unique as can be seen in the admin interface:

<img width="1315" alt="image" src="https://user-images.githubusercontent.com/1621608/225501157-b7600ac2-e907-4203-a3ee-ae4b41c7e884.png">

This constraint can result in failing API requests when attempting to create a data stream that has a non-unique name

<img width="1509" alt="image" src="https://user-images.githubusercontent.com/1621608/225501556-cba43df6-4a74-41ba-99bb-6dfa422fa0d1.png">

While data streams are not the GA4 equivalent of UA views, this is somewhat related to the work we did for enhancing the creation of a new view (see https://github.com/google/site-kit-wp/issues/716).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation Brief

* 

### Test Coverage

* 

## QA Brief

* 

## Changelog entry

*

**Comment by wpdarren** [comment]
I would like to include two bugs I have found while testing #6330 that I understand will be fixed in this ticket. 

A console error and analytics is not successfully set up in a specific scenario:  I have a site that has Site Kit and Analytics already set up with a web data steam. I reset and set up again. Rather than go with the defaulted web data stream, I select from the dropdown to create a new one. When clicking on the configure analytics CTA, I am redirected back to the dashboard and the success notification appears. 

In the console though an error message appears on clicking the CTA. 

`Google Site Kit API Error method:POST datapoint:create-webdatastream type:modules identifier:analytics-4 error:"Requested entity already exists"  POST https://ing-online-voka.instawp.xyz/wp-json/google-site-kit/v1/modules/analytics-4/data/create-webdatastream?_locale=user 409`

When I go to the settings for Analytics, the module isn't connected and when editing the settings the GA4 property and web data stream fields are blank.   You can see it in action on this screencast. 

https://user-images.githubusercontent.com/73545194/227140062-c956843c-8bd5-408c-a6fa-428591a1b892.mp4

A similar issue occurs when I go to edit settings. I select a new web data stream, and click on the confirm changes and the settings are saved but no new web data stream is created and the same error above appears. 

https://user-images.githubusercontent.com/73545194/227141021-8ff82f6f-9a28-4f8f-8d20-14e4d59541db.mp4

@aaemnnosttv since #6330 is going out in 1.98.0, do we have any plans to include this fix in the same release? Otherwise we will have a bug which I feel will cause support tickets.

**Comment by aaemnnosttv** [comment]
@wpdarren it's not really a bug on our end but not a great experience for the user, I agree.

Even though it's not too hard to reproduce, it should be a rare edge case. Even if we did get a request about it, there is an easy work around to simply rename the existing data stream and try again. Ideally we'd address it in the upcoming release, but it needs to wait until all the high priority issues for GA4 reporting are done.

**Comment by eugene-manuilov** [comment]
@zutigrm 

> * In `assets/js/modules/analytics-4/datastore/settings.js`:
>   
>   * Pass the `webDataStreamName` to the `createWebDataStream` [action](https://github.com/google/site-kit-wp/blob/996e3acf35f9b42e1da410f96b8aadffa6ece18a/assets/js/modules/analytics-4/datastore/settings.js#L92)

We also need to update the `validateCanSubmitChanges` function to verify that the web data stream name is unique: https://github.com/google/site-kit-wp/blob/996e3acf35f9b42e1da410f96b8aadffa6ece18a/assets/js/modules/analytics-4/datastore/settings.js#L189

> * Update `assets/js/modules/analytics-4/components/common/WebDataStreamSelect.js`:
>   
>   * Render new `WebDataStreamNameInput` component bellow the select, if `WEBDATASTREAM_CREATE` is selected

It should be rendered in the same way how we render the profile name field: https://github.com/google/site-kit-wp/blob/996e3acf35f9b42e1da410f96b8aadffa6ece18a/assets/js/modules/analytics/components/setup/SetupFormUA.js#L102-L106

In other words, we need to render this component on the next line below all 3 components (account/property/webdatastream selects). It means that we need to update setup and settings forms to have the same behavior on all of them.

> Add `assets/js/modules/analytics-4/components/common/WebDataStreamNameInput.js`

This component also should display the create web data stream error instead of doing it in the web data stream select if we create a new data stream.

**Comment by zutigrm** [comment]
@eugene-manuilov Thank you for the feedback. IB updated

**Comment by eugene-manuilov** [comment]
Nice. Thanks, @zutigrm. IB ✔️

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on dev environment.
- Verified web data stream creation and validation when creating new analytics a/c during Analytics set up.
- Verified new web data stream creation and validation when connecting with existing analytics a/c during Analytics set up.
- Verified new web data stream creation and validation under Analytics settings.
- Verified that validation messages are appearing and user is not able to complete analytics setup while creating new analytics a/c or when trying to connect with existing analytics a/c.

**Note :** User is able to save settings even if user enter already exist web data stream name or do not provide any web data stream name which is known issue as per QAB. I created separate issue for this - https://github.com/google/site-kit-wp/issues/8467

_1) Verified web data stream creation and validation when creating new analytics a/c during Analytics set up._

<details>

https://github.com/google/site-kit-wp/assets/94359491/aa999b64-e62c-4698-be57-6916ec0a92a9

</details>

_2) Verified new web data stream creation and validation when connecting with existing analytics a/c during Analytics set up._

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/6b31d246-111c-4431-950f-2e7b48d39081)

![image](https://github.com/google/site-kit-wp/assets/94359491/1e0def82-3966-4502-967f-deda822a961a)

https://github.com/google/site-kit-wp/assets/94359491/7a3d87c0-a5cb-43d0-9f61-d7a1349ad82d

</details>

_3) Verified new web data stream creation and validation under Analytics settings._

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/2bcf2374-50a3-40e7-aba6-714d393139e6)

![image](https://github.com/google/site-kit-wp/assets/94359491/85f27e1f-c220-4374-aaaf-af7a5c97c875)

https://github.com/google/site-kit-wp/assets/94359491/fc09bf4d-9eee-4d32-a552-8c77ad5adc82

</details>

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

Moving this back to QA from approval because I have one question. 

@nfmohit I noticed that when user save analytics settings after giving invalid web data stream name or without providing web data stream name then settings gets save and analytics setup status set to incomplete. Now when again user trying to complete setup then user is not able to complete setup even after providing unique web data stream name. In this case complete setup button remains inactive until user selects already exists web data stream. So, in this case user will not able to create new web data stream. Can you implement fix for this or this scenario will be address under https://github.com/google/site-kit-wp/issues/8467 which I created to address settings allow saving of invalid web data stream name.


https://github.com/google/site-kit-wp/assets/94359491/2d558e12-c0ce-4d1a-832c-1a7e954fbe25

**Comment by nfmohit** [comment]
Hi @mohitwp. I have actually noticed this problem once before. This problem has more to do with the submit settings button ("Confirm Changes"/"Save") rather than the web data stream name input.

Currently, if there are no changes in the settings, or there is an invalid setting value, the button simply says "Save", whereas, it should keep saying "Confirm Changes" but just be disabled, so that the user cannot submit.

For example, to understand this behaviour, try to edit settings of Tag Manager. When you're there, choose to create a new container, but keep the field blank. Notice how the submit button says "Save" and you can still submit, where you shouldn't be able to. Once you click "Save", it just gets stuck at the loading state and never completes.

As this issue doesn't seem specific to the web data stream input but the settings edit form for modules in general, do you think it might be wiser to create a new issue for further investigation?

Thanks!

**Comment by mohitwp** [comment]
### QA Update ✅ 

As per slack discussion we will address [above](https://github.com/google/site-kit-wp/issues/6727#issuecomment-2034108343) issue through #8467 

- Tested on dev environment.
- Verified web data stream creation and validation when creating new analytics a/c during Analytics set up.
- Verified new web data stream creation and validation when connecting with existing analytics a/c during Analytics set up.
- Verified new web data stream creation and validation under Analytics settings.
- Verified that validation messages are appearing and user is not able to complete analytics setup while creating new analytics a/c or when trying to connect with existing analytics a/c.

**Note :** User is able to save settings even if user enter already exist web data stream name or do not provide any web data stream name which is known issue as per QAB. I created separate issue for this - https://github.com/google/site-kit-wp/issues/8467

_1) Verified web data stream creation and validation when creating new analytics a/c during Analytics set up._

<details>

https://github.com/google/site-kit-wp/assets/94359491/aa999b64-e62c-4698-be57-6916ec0a92a9

</details>

_2) Verified new web data stream creation and validation when connecting with existing analytics a/c during Analytics set up._

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/6b31d246-111c-4431-950f-2e7b48d39081)

![image](https://github.com/google/site-kit-wp/assets/94359491/1e0def82-3966-4502-967f-deda822a961a)

https://github.com/google/site-kit-wp/assets/94359491/7a3d87c0-a5cb-43d0-9f61-d7a1349ad82d

</details>

_3) Verified new web data stream creation and validation under Analytics settings._

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/2bcf2374-50a3-40e7-aba6-714d393139e6)

![image](https://github.com/google/site-kit-wp/assets/94359491/85f27e1f-c220-4374-aaaf-af7a5c97c875)

https://github.com/google/site-kit-wp/assets/94359491/fc09bf4d-9eee-4d32-a552-8c77ad5adc82

</details>

**Comment by mohitwp** [comment]
### QA Update ❌ 

Moving this back to execution due to regression issue found during release QA.

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on main environment..
- Verified Analytics set up.
- Verified create a/c , Create new property and create new Web data stream functionality.
- Now user is able to create new property.


https://github.com/google/site-kit-wp/assets/94359491/0fd3b669-49f3-4e2e-8468-1a1a30cc4c41

**Comment by aaemnnosttv** [comment]
## ⚠️ Approval

There is an odd experience when switching between properties as can be seen in @mohitwp 's screencast above where the field is visible while the dropdowns are loading. This should be fixed.

<img width="605" alt="image" src="https://github.com/google/site-kit-wp/assets/1621608/11e21924-9d00-450a-a804-ff5f603fd41a">

I went back and looked at how we did this in the legacy Analytics and it seems about the same, but we handle the process of selecting a property a bit differently. I'll see about putting together a quick PR for this.

**Comment by nfmohit** [comment]
The above issue has now been addressed through [a follow-up PR](https://github.com/google/site-kit-wp/pull/8490). This is back in QA.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- Went through switching between different Analytics accounts, properties and web data streams.
- Went through creating new accounts, properties and web data streams. 
- Didn't find any regressions, no issues with the odd UX/UI as observed by Evan [here](https://github.com/google/site-kit-wp/issues/6727#issuecomment-2040483849).
